src/crystal/hasher.cr

```diff
@@ -45,12 +74,12 @@ struct Crystal::Hasher
   end
 
   def reference(value)
-    @result = @result * 31 + value.object_id.to_u64
-    self
+    value.object_id.to_u64.hash(self)
   end
 
   def string(value)
-    value.to_slice.hash(self)
+    @siphash.update(value)
```

Comment:
Breaking change: `"some data".hash` and `"some data".to_slice.hash` no longer have the same hash!

- `String#hash` will compress 4-bytes at the time;
- `Slice#hash` will iterate each byte, casted to UInt64, and finally compressed (as 8 bytes).

Reply:
> `Slice#hash` will iterate each byte, casted to UInt64, and finally compressed (as 8 bytes).

Is there a performance _penalty_ to do it byte by byte?

Reply:
Yes. SipHash will only compress 8 plain bytes at once (4 bytes with halfsiphash) —no more, no less, unless we finalize — if it's smaller or not a multiple of 8 bytes (or 4 bytes with halfsiphash) then we must buffer bytes until we have enough to compress —it adds to the overhead.

Furthermore, encoding `"some data"` will result in compressing `"some dat"` then `"a"` (thus call `update` once then finalize), whereas `"some data".to_slice` will call `update` 9 times (for each byte casted to UInt64).

Note that there is little we can do: Slice(T) may be any T, and we must properly call `T#hash(hasher)` for each entry.

Reply:
I don't quite think this is a "real" breaking change, as `"some data" != "some data".to_slice`. People can't depend on the output of `hash` even after 1.0. They can only depend on the invariants around `==`.

Reply:
Thanks for the explanation
> Note that there is little we can do: `Slice(T)` may be any T, and we must properly call `T#hash(hasher)` for each entry.

You're right, maybe we could (in an other PR?) add an optimization for some `T` types, like `Bytes` (`Slice(UInt8)`)? And/Or more generally to any T that is a `Number` (and `Bool`)?

Reply:
`Slice(Primitive)` is very common case and should be optimized anyway.